overlord/hookstate/hookstate_test.go

```diff
@@ -652,3 +675,137 @@ func (s *hookManagerSuite) TestHookTaskHandlerReportsErrorIfRequested(c *C) {
 
 	c.Check(errtrackerCalled, Equals, true)
 }
+
+func (s *hookManagerSuite) TestHookTasksForSameSnapAreSerialized(c *C) {
+	hooksup := &hookstate.HookSetup{
+		Snap:     "test-snap",
+		Hook:     "configure",
+		Revision: snap.R(1),
+	}
+
+	s.state.Lock()
+
+	var tasks []*state.Task
+	for i := 0; i < 20; i++ {
+		task := hookstate.HookTask(s.state, "test summary", hooksup, nil)
+		c.Assert(s.task, NotNil)
+		change := s.state.NewChange("kind", "summary")
+		change.AddTask(task)
+		tasks = append(tasks, task)
+	}
+	s.state.Unlock()
+
+	s.settle()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	c.Check(s.task.Kind(), Equals, "run-hook")
+	c.Check(s.task.Status(), Equals, state.DoneStatus)
+	c.Check(s.change.Status(), Equals, state.DoneStatus)
+
+	for i := 0; i < len(tasks); i++ {
+		c.Check(tasks[i].Kind(), Equals, "run-hook")
+		c.Check(tasks[i].Status(), Equals, state.DoneStatus)
+	}
+	c.Assert(atomic.LoadInt32(&s.mockHandler.TotalExecutions), Equals, int32(1+len(tasks)))
+	c.Assert(atomic.LoadInt32(&s.mockHandler.Executing), Equals, int32(0))
+}
+
+type MockConcurrentHandler struct {
+	onDone func()
+}
+
+func (h *MockConcurrentHandler) Before() error {
+	return nil
+}
+
+func (h *MockConcurrentHandler) Done() error {
+	h.onDone()
+	return nil
+}
+
+func (h *MockConcurrentHandler) Error(err error) error {
+	return nil
+}
+
+func NewMockConcurrentHandler(onDone func()) *MockConcurrentHandler {
+	return &MockConcurrentHandler{onDone: onDone}
+}
+
+func (s *hookManagerSuite) TestHookTasksForDifferentSnapsRunConcurrently(c *C) {
+	hooksup1 := &hookstate.HookSetup{
+		Snap:     "test-snap-1",
+		Hook:     "prepare-device",
+		Revision: snap.R(1),
+	}
+	hooksup2 := &hookstate.HookSetup{
+		Snap:     "test-snap-2",
+		Hook:     "prepare-device",
+		Revision: snap.R(1),
+	}
+
+	s.state.Lock()
+
+	sideInfo := &snap.SideInfo{RealName: "test-snap-1", SnapID: "some-snap-id1", Revision: snap.R(1)}
+	info := snaptest.MockSnap(c, snapYaml1, snapContents, sideInfo)
+	c.Assert(info.Hooks, HasLen, 1)
+	snapstate.Set(s.state, "test-snap-1", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{sideInfo},
+		Current:  snap.R(1),
+	})
+
+	sideInfo = &snap.SideInfo{RealName: "test-snap-2", SnapID: "some-snap-id2", Revision: snap.R(1)}
+	snaptest.MockSnap(c, snapYaml2, snapContents, sideInfo)
+	snapstate.Set(s.state, "test-snap-2", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{sideInfo},
+		Current:  snap.R(1),
+	})
+
+	var testSnap1HookCalls, testSnap2HookCalls int
+	ch := make(chan struct{}, 0)
```

Comment:
@stolowski thanks, the langref is not super clear to what make(chan T, 0) 

https://golang.org/ref/spec#Making_slices_maps_and_channels

it seems you get un unbuffered channel, but a buffered channel where send would always fail would seem not ruled out. it would be  bit useless, though that's the behavior of the nil channel